dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/search/SolrQuerySanitizer.java

```diff
@@ -0,0 +1,143 @@
+    
+    
+    @Inject
+    private DatasetFieldServiceBean datasetFieldService;
+    @Inject
+    private SolrFieldFactory solrFieldFactory;
+    
+    // -------------------- LOGIC --------------------
+    
+    public String sanitizeQuery(String query) {
+        if (query == null) {
+            return StringUtils.EMPTY;
+        }
+        
+        query = replaceQueryFieldNames(query, buildFieldNamesMapping());
+        query = escapeGlobalIdValues(query);
+        
+        return query;
+    }
+    
+    // -------------------- PRIVATE --------------------
+    
+    private Map<String, String> buildFieldNamesMapping() {
+        Map<String, String> fieldNamesMapping = new HashMap<>();
+        List<DatasetFieldType> datasetFields = datasetFieldService.findAllOrderedById();
+        for (DatasetFieldType datasetFieldType : datasetFields) {
+
+            SolrField dsfSolrField = solrFieldFactory.getSolrField(datasetFieldType.getName(),
+                                                                   datasetFieldType.getFieldType(),
+                                                                   datasetFieldType.isThisOrParentAllowsMultipleValues(),
+                                                                   datasetFieldType.isFacetable());
+            fieldNamesMapping.put(datasetFieldType.getName(), dsfSolrField.getNameSearchable());
+        }
+        return fieldNamesMapping;
+    }
+    
+    private void transferString(StringBuilder source, StringBuilder target) {
+        target.append(source);
+        source.setLength(0);
+    }
+    
+    private String replaceQueryFieldNames(String query, Map<String, String> fieldNamesMapping) {
+        StringBuilder transformedQuery = new StringBuilder();
+
+        StringBuilder currentSegmentWithoutSpecialCharacters = new StringBuilder();
+
+        boolean insideQuotation = false;
+        boolean currentCharIsEscaped = false;
+        
+        for (int i=0; i<query.length(); ++i) {
+            char currentChar = query.charAt(i);
+
+            if (currentChar == '\\') {
+                transformedQuery.append(currentChar);
+                currentCharIsEscaped = !currentCharIsEscaped;
+                continue;
+            }
+            
+            if (currentChar == '"' && !currentCharIsEscaped) {
+                insideQuotation = !insideQuotation;
+
+                if(insideQuotation) {
+                    transferString(currentSegmentWithoutSpecialCharacters, transformedQuery);
+                }
+            }
+
+            if (insideQuotation) {
+                transformedQuery.append(currentChar);
+                continue;
+            }
+
+            if (currentChar == ':') {
+                String fieldName = currentSegmentWithoutSpecialCharacters.toString();
+                
+                if (fieldNamesMapping.containsKey(fieldName)) {
+                    fieldName = fieldNamesMapping.get(fieldName);
```

Comment:
And here solrFieldName, I know the logic will be kinda "bigger" and you would have to use ternary operator when appending, but it could be more explanatory.

Reply:
You mean:
```
            if (currentChar == ':') {
                String fieldNameForView = currentSegmentWithoutSpecialCharacters.toString();
                
                String solrFieldName = fieldNamesMapping.containsKey(fieldNameForView) ? 
                        fieldNamesMapping.get(fieldNameForView) : fieldNameForView;
                
                transformedQuery.append(solrFieldName);
                currentSegmentWithoutSpecialCharacters.setLength(0);
                transformedQuery.append(currentChar);
                continue;
            }
```
?

Reply:
Something more like :
```
if (currentChar == ':') {
                String fieldNameFromView = currentSegmentWithoutSpecialCharacters.toString();
                Option<String> solrFieldName = Option.none();
                
                if (fieldNamesMapping.containsKey(fieldNameFromView)) {
                    solrFieldName = Option.of(fieldNamesMapping.get(fieldNameFromView));
                }
                
                transformedQuery.append(solrFieldName.getOrElse(fieldNameFromView));
                currentSegmentWithoutSpecialCharacters.setLength(0);
                transformedQuery.append(currentChar);
                continue;
            }
```